# Coding Exercises (Part 1)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Pandas: Basics

### Exercise 1: Importing and Inspecting your own Dataset

Before we can work with the Pandas Library, we need to __import Pandas__ (by convention as __pd__). <br>
1. __Import__ Pandas. __Fill in the gaps__!

In [2]:
import pandas as pd

Now we want to import the __Listings Dataset__ that is stored in the CSV-File __listings.csv__. <br>
2. __Import__ the Listings Dataset and __store__ the Pandas DataFrame in the variable __listings__! __Fill in the gaps__!

In [9]:
listings = pd.read_csv("listings.csv")

#### Some more details on the Columns:

- __Symbol__: A Stock Ticker Symbol is an abbreviation used to uniquely identify publicly traded shares of a particular stock on a particular stock market
- __Exchange__: Marketplace in which shares are traded 
- __Name__: Legal Company Name
- __Last_Price__: Last Trading Price (as of April 2019)
- __Market_Cap__: Dollar Value of Outstanding Shares (as of April 2019). Computed as Shares times current Price. 
- __ADR_TSO__: Additional Information on foreign stocks trading in the US
- __IPO_Year__: Year of Initial Public Offering 
- __Sector__: Sector of main business activity
- __Industry__: Industry of main business activity

3. __Inspect__ the __first 10 Rows__ of the DataFrame listings! __Fill in the gaps__! <br> What is the __company name of the very first listing__ in the first row (@ index label/pos = 0)?

In [11]:
listings.iloc[0,2]

'Agilent Technologies, Inc.'

It´s ... Agilent Technologies, Inc.!

4. Also __inspect__ the __last 5 Rows__! __How many listings/rows__ do we have in the Dataset (mind the range index starting from 0!)?

In [15]:
listings.tail(5)

Symbol Exchange                           Name  Last_Price    Market_Cap  \
6847   ZUMZ   NASDAQ                    Zumiez Inc.       26.72  6.819191e+08   
6848    ZUO     NYSE                    Zuora, Inc.       19.79  2.147266e+09   
6849   ZYME     NYSE                 Zymeworks Inc.       15.74  5.040782e+08   
6850   ZYNE   NASDAQ  Zynerba Pharmaceuticals, Inc.        7.85  1.653995e+08   
6851   ZYXI   NASDAQ                    Zynex, Inc.        5.02  1.618349e+08   

      ADR TSO  IPO_Year             Sector  \
6847      NaN    2005.0  Consumer Services   
6848      NaN    2018.0         Technology   
6849      NaN    2017.0        Health Care   
6850      NaN    2015.0        Health Care   
6851      NaN       NaN        Health Care   

                                               Industry  
6847                     Clothing/Shoe/Accessory Stores  
6848            Computer Software: Prepackaged Software  
6849                              Major Pharmaceuticals  
6850                              Major Pharmaceuticals  
6851  Biotechnology: Electromedical & Electrotherape...

We have ... 6852 listings!

5. Get some __meta information__ on our DataFrame! In which __Column(s)__ do we have __missing/NaN__ Values?

In [16]:
listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      6852 non-null   object 
 1   Exchange    6852 non-null   object 
 2   Name        6852 non-null   object 
 3   Last_Price  6745 non-null   float64
 4   Market_Cap  5954 non-null   float64
 5   ADR TSO     140 non-null    float64
 6   IPO_Year    3105 non-null   float64
 7   Sector      5309 non-null   object 
 8   Industry    5309 non-null   object 
dtypes: float64(4), object(5)
memory usage: 481.9+ KB


We have missing/NaN Values in the columns ... Last_Price, Market_Cap, ADR TSO, IPO_Year, Sector & Industry!

6. Let´s get some __summary statistics__ on our DataFrame! What is the __maximum__ value in the column __Last_Price__?

In [17]:
listings.describe()

Last_Price    Market_Cap       ADR TSO     IPO_Year
count  6745.000000  5.954000e+03  1.400000e+02  3105.000000
mean     37.372232  7.174440e+09  5.869242e+07  2009.025765
std      99.249040  3.525455e+10  1.629909e+08     9.337562
min       0.003000  0.000000e+00  2.400000e+04  1972.000000
25%       8.310000  1.203769e+08  4.271224e+06  2003.000000
50%      20.573500  5.585176e+08  1.406964e+07  2013.000000
75%      37.370000  2.888747e+09  3.962613e+07  2017.000000
max    4299.990000  9.459795e+11  1.485426e+09  2019.000000

The maximum value in the column Last_Price is ... 4299.99!

# Well Done!

---------------------------------------------------------------------------------------------------------------------

# Hints (Spoiler!)

1. pandas as pd

2. Pass the filename in quotation marks to pd.read_csv()

3. head() method with n = 10

4. tail() method

5. info() method

6. describe() method